doc: publish the graphics enablement doc on rtd (#3334)
canonical/mir: doc/sphinx/how-to/how-to-enable-graphics-core22-on-a-device.md

    override-build: |
      snapcraftctl build
      # We use the libwayland-egl from the vendor-drivers part.
      # We remove this here, as otherwise snapcraft will complain that we have two
      # different files trying to provide libwayland-egl.so
      rm $SNAPCRAFT_PART_INSTALL/usr/local/lib/aarch64-linux-gnu/libwayland-egl*
    source: https://gitlab.freedesktop.org/wayland/wayland.git
    source-type: git
    source-tag: 1.21.0
    organize:
      'usr/local/lib/aarch64-linux-gnu': graphics/lib/
    prime:
      - graphics/lib/*
```

#### Vendor driver uses a $VENDOR_dri.so DRI driver
Some vendor drivers (internally based on Mesa) retain the separate DRI driver component. If
a vendor driver is missing its DRI component then running a program using EGL will generate
a `libGL` warning about failure to load `$VENDOR_dri.so`, followed by either program failure
or possibly the program continuing with CPU rendering.

Any vendor DRI driver should be provided under the top-level dri/ directory. An example
from the `imx-gpu-viv` driver is:

```yaml
# TODO: What is the context here?
organize:
  imx-gpu-viv-6.4.3.p0.0-aarch64/gpu-core/usr/lib/dri/* : graphics/dri/

> ```

Author: @RAOF

> ```

Reviewer: This is part of a `vendor-drivers` part.

> ```

Author: You mean `dependencies` in the previous example, right?

Snaps which use the `graphics-core22` interface will set the necessary environment variables
to point the DRI loader at the top-level `dri/` path under the path exported by the
`graphics-core22` slot.

#### Vendor drivers use libglvnd
Some drivers might use the `libglvnd` vendor-neutral GL/EGL library. In this case the vendor
drivers will ship an `egl_vendor.d/$SOMETHING.json` file, pointing to their vendor EGL library.

> The [mesa-core22](https://github.com/canonical/mesa-core22/blob/main/snap/snapcraft.yaml)

Author: @RAOF

> The [mesa-core22](https://github.com/canonical/mesa-core22/blob/main/snap/snapcraft.yaml)

Reviewer: Actually, I think this is now exported in `share/egl_vendor.d`

> The [mesa-core22](https://github.com/canonical/mesa-core22/blob/main/snap/snapcraft.yaml)

Author: It is actually in `usr/share/glvnd`

# How to enable graphics-core22 on a device

This document will guide you through the process of assembling a 
`graphics-core22` snap. With this snap, your device will be ready
to run graphical snaps like `ubuntu-frame` or any other *Mir*-based
compositor in a snapped environment.

To this end, we will accomplish the following in order:
- How to assemble a `graphics-core22` provider snap
- How to test a `graphics-core22` provider snap
- How to test the confinement of your *Mir*-based snap that relies on the provider snap

## Prerequisites
Before we get started, let's make sure that we have everything that we require.

### 1. Development Board
For starters, make sure that you have some sort of development board on hand.

### 2. GPU Support Check
Next, make sure that this board runs a `snapd`-enabled Ubuntu image with
a kernel that includes GPU support. To do this, we'll run a series of checks.

First, you should ensure that *at least* `/dev/dri/card0` exists, but you 
may have more depending on your setup. You may check this from the command line.

Next, install [drm_info](https://gitlab.freedesktop.org/emersion/drm_info). The

> output of `drm_info` will look something like the following:

Reviewer: This is available in the `graphics-test-tools` snap, which *can* be installed before having a graphics-core22 provider (it'll pull in the mesa provider)

```
Node: /dev/dri/card1
├───Driver: amdgpu (AMD GPU) version 3.54.0 (20150101)
│   ├─── ...
├───Device: PCI 1002:747e Advanced Micro Devices, Inc. [AMD/ATI] Device 747e
│   └───Available nodes: primary, render
├───Framebuffer size
│   ├───Width: [0, 16384]
│   └───Height: [0, 16384]
├───Connectors
|   |   ...
│   ├───Connector 3
│   ├───Object ID: 101
│   ├───Type: HDMI-A
│   ├───Status: connected
│   ├───Physical size: 530x300 mm
│   ├───Subpixel: unknown
│   ├───Encoders: {3}
│   ├───Modes
│   │   ├───1920x1080@60.00 preferred driver phsync pvsync 
│   │   ├───1920x1080@60.00 driver phsync pvsync 16:9 
│   │   ├───...
|   |   ...
├───Encoders
│   ├───Encoder 0
│   │   ├───Object ID: 82
│   │   ├───Type: TMDS
│   │   ├───CRTCS: {0, 1, 2, 3}
│   │   └───Clones: {0}
|   |   ...
├───CRTCs
│   ├───CRTC 0
│   │   ├───Object ID: 72
│   │   ├───Legacy info
│   │   │   ├───Mode: 1920x1080@60.00 preferred driver phsync pvsync 
│   │   │   └───Gamma size: 256
|   |   ...
└───Planes
    ├───Plane 0
    │   ├───Object ID: 40
    │   ├───CRTCs: {3}
    │   ├───Legacy info
    │   │   ├───FB ID: 0
    |   ...
```
You want to make sure that you have:

> - 1 DRM Node (?) TOOD

Reviewer: At least one.

(What's the TOOD? :wink:)

Author: OH! I meant to ask you here. it worked! 

What do you mean by "1 DRM Node" in the output? I don't see any listed there

Author: We meant `Node` here as the "drm node"

- 1 or more Connector(s)
- 1 or mor CRTC(s)
- 1 or more Plane(s)

Finally, If you have a display connected, then one of your `Connector`s should read `Status: connected`. This `Connector` will also list many `Modes` that
the  display can use.

#### Troubleshooting
If `drm_info` does *not* produce results similar to those above, then this
indicates that your kernel **is NOT yet ready** to support the GPU. This

> means that kernel enablement is required before ubuntu frame can be enabled.

Reviewer: ```suggestion
means that kernel enablement is required before Ubuntu Frame can be enabled.
```

### 3. Driver Support Check
Finally, we will need to check that you have the proper userspace drivers
installed on the system, namely:
- `EGL` (`libEGL.so.1`)
- `GLES` (`libGLESv2.so.2`) 
- `GBM` (`libgbm.so.1`)

While *Mir* can be enabled on platforms that do not support `GBM/KMS`, this
requires a paltform implementation to be written in *Mir*. This work is out
of the scope of this guide.

## How to assemble a `graphics-core22` provider snap
With the prerequisites out of the way, it is time to assemble a
`graphics-core22` snap.

The snap itself must provide a handful libraries
which fall into the following four categories:

- Graphics interfaces
  - `libEGL.so.1`
  - `libva.so.2`
  - `libvulkan.so.1`
  - `libGLESv2.so.2`

- GBM/Kms interfaces
  - `libdrm.so.2`
  - `libgbm.so.1`

- Wayland support
  - `libwayland-client.so.0`
  - `libwayland-server.so.0`

- X support (optional)
  - `libX11-xcb.so.1`
  - `libXau.so.6`
  - `libxcb-dri2.so.0`
  - `libxcb-dri3.so.0`
  - `libxcb-present.so.0`
  - `libxcb.so.1`
  - `libxcb-sync.so.1`
  - `libxcb-xfixes.so.0`
  - `libXdmcp.so.6`
  - `libxshmfence.so.1`

> `libEGL`, `libGLESv2` and `libgbm` should be provided by the vendors drivers.

Reviewer: ```suggestion
`libEGL`, `libGLESv2` and `libgbm` should be provided by the vendor's drivers.
```

Reviewer: Also, `libvulkan` *may* be provided by the vendor. I'm not entirely sure how best to word that.

Author: We could mentioned that it _may_ be provided. I think that at least prompts the user to check in some way.

or possibly the program continuing with CPU rendering.

Any vendor DRI driver should be provided under the top-level dri/ directory. An example
from the `imx-gpu-viv` driver is:

```yaml
# TODO: What is the context here?
organize:
  imx-gpu-viv-6.4.3.p0.0-aarch64/gpu-core/usr/lib/dri/* : graphics/dri/
```

Snaps which use the `graphics-core22` interface will set the necessary environment variables
to point the DRI loader at the top-level `dri/` path under the path exported by the
`graphics-core22` slot.

#### Vendor drivers use libglvnd
Some drivers might use the `libglvnd` vendor-neutral GL/EGL library. In this case the vendor
drivers will ship an `egl_vendor.d/$SOMETHING.json` file, pointing to their vendor EGL library.
The [mesa-core22](https://github.com/canonical/mesa-core22/blob/main/snap/snapcraft.yaml)
`snapcraft.yaml` provides an example of this sort of driver; the `egl_vendor.d`
directory is shipped in a top-level `glvnd` path, under the path exported by the
`graphics-core22` slot.

## How to test a `graphics-core22` provider snap
Now we have our `graphics-core22` provider snap built. However, we don't yet know if it works
properly. Testing this snap is our next task.

> For starters, install the [graphics-test-tools](https://snapcraft.io/graphics-test-tools) snap:

Reviewer: We should probably move this bit up to the top, so that they have `drm-info` early.

```shell
sudo snap install graphics-test-tools --channel 22/stable
```

Next, connect the snap that you've built to `graphics-test-tools`:

```shell
sudo snap connect graphics-test-tools:graphics-core22 <your-snap>:graphics-core22
```

With that installed, we can begin testing.

### Use `eglinfo` to test
This is the most basic test that can be performed, and is a good indication of baseline GPU
driver setup. This should list _at least_ a GBM platform with the expected vendor and 

> `OpenGL_ES` client API. An abbreviated example (on a `mesa-core22` system):

Author: @RAOF I changed this from `mesa-core20` to `mesa-core22`. Is that a problem?

Reviewer: No.

First, install `ubuntu-frame`:
```shell
sudo snap install ubuntu-frame --devmode
```

Next, disconnect it from the default graphics interface:
```shell
sudo snap disconnect ubuntu-frame:graphics-core22
```

Then, connect it to your new provider snap:
```shell
sudo snap connect ubuntu-frame:graphics-core22 <your-snap>:graphics-core22
```

Finally, run `ubuntu-frame`:
```shell
ubuntu-frame
```

If everything previously has gone correctly this should result in `ubuntu-frame` coming
up on the connected outputs. If not, the log messages will hopefully help identify issues.

## How to test the confinement of your *Mir*-based snap
Once you have your `graphics-core22` provider snap setup, and you are confident that it works
well, you'll want to make sure that you can run the snap that *depends on* your provider snap
in a confined mode. Since we already have `ubuntu-frame` setup, we will use it as an example.
However, you should be able to run these tests on any *Mir*-based snap.

First, let's uninstall `ubuntu-frame` if it is installed:

```shell
sudo snap remove ubuntu-frame
```

Next, we can install `ubuntu-frame` without the `--devmode` flag:
```shell
sudo snap install ubuntu-frame

> sudo snap disconnect ubuntu-frame:graphics-core22

Reviewer: We also need to *connect* to the `graphics-core22` slot of the driver provider.

```

Finally, we will run `ubuntu-frame` like before:
```shell
ubuntu-frame
```

If everything works, then `ubuntu-frame` is ready. If not, we'll have to troubleshoot.

### Troubleshooting
When bringing up a confined ubuntu-frame snap on a new board with new drivers there are two
separate access control mechanisms:
- AppArmor
- The devices cgroup

#### AppArmor
The first is relatively easy to debug. When AppArmor would deny access to a resource, it outputs a nice
message in `dmesg`:

```
apparmor="DENIED" operation="open" profile="snap.ubuntu-frame.ubuntu-frame" name="/run/udev/data/c189:1" <other stuff>
```

You can find the AppArmor profile at `/var/lib/snapd/apparmor/profiles/snap.ubuntu-frame.ubuntu-frame`.
Make modifications to the profile using your favorite text editor. Finally, apply the change to replace
the default confinement:

```shell
sudo apparmor_parser -r /var/lib/snapd/apparmor/profiles/snap.ubuntu-frame.ubuntu-frame

> ```

Reviewer: Probably worth adding a bit about how this profile is regenerated on lots of `snapd` actions, so these modifications are *temporary* and will need to be upstreamed via a bug against snapd to become useful.

a vendor driver is missing its DRI component then running a program using EGL will generate
a `libGL` warning about failure to load `$VENDOR_dri.so`, followed by either program failure
or possibly the program continuing with CPU rendering.

Any vendor DRI driver should be provided under the top-level dri/ directory. An example
from the `imx-gpu-viv` driver is:

```yaml
parts:
  ...
  vendor-drivers:
    ...
    organize:
      imx-gpu-viv-6.4.3.p0.0-aarch64/gpu-core/usr/lib/dri/* : graphics/dri/
```

Snaps which use the `graphics-core22` interface will set the necessary environment variables
to point the DRI loader at the top-level `dri/` path under the path exported by the
`graphics-core22` slot.

#### Vendor drivers use libglvnd
Some drivers might use the `libglvnd` vendor-neutral GL/EGL library. In this case the vendor
drivers will ship an `egl_vendor.d/$SOMETHING.json` file, pointing to their vendor EGL library.
The [mesa-core22](https://github.com/canonical/mesa-core22/blob/main/snap/snapcraft.yaml)
`snapcraft.yaml` provides an example of this sort of driver; the `egl_vendor.d`
directory is shipped in a top-level `usr/share/glvnd` path, under the path exported by the

> `graphics-core22` slot.

Reviewer: This whole bit needs to change significantly, sorry.

We need to make clear a couple of things:
* The snap should provide as many of [the supported ABIs](https://mir-server.io/docs/the-graphics-core22-snap-interface#heading--lists--supported-apis) as possible.
* Those libraries don't need to be in any particular location.
* There's only a single file that the provider *does* need to have in a specific location: `bin/graphics-core22-provider-wrapper`, which gets called as `graphics-core22-provider-wrapper $SOME_BINARY $ARGS...` and needs to set up the environment so that `$SOME_BINARY` will find the libraries and then execute `$SOME_BINARY` with `$ARGS...`.

We should probably do a simple example (a libmali provider?) and then follow that up with an annotated description of the [`mesa-core22` script](https://github.com/canonical/mesa-core22/blob/main/scripts/bin/graphics-core22-provider-wrapper.in) that covers the `DRI_DRIVER_PATH` and glvnd issues.

While *Mir* can be enabled on platforms that do not support `GBM/KMS`, this
requires a paltform implementation to be written in *Mir*. This work is out
of the scope of this guide.

## How to assemble a `graphics-core22` provider snap
With the prerequisites out of the way, it is time to assemble a `graphics-core22` snap.

The snap itself must provide the libraries described in 
[this document](https://mir-server.io/docs/the-graphics-core22-snap-interface#heading--lists--supported-apis).
Unless you are incredibly space-constrained or in an unusual circumstance, it is a good idea to
provide **all** of these libraries. Even if your board doesn’t support these libraries, a failure to include
them will result in applications that would otherwise fall back to a supported API failing to start.

`libEGL`, `libGLESv2` and `libgbm` should be provided by the vendor's drivers.
`libvulkan` *may* be provided by the vendor as well.
The remaining packages should be available in the Ubuntu archive.

Most of these libraries do **NOT** need to be in a fixed location in the snap.
However, the following paths **MUST** be provided in a fixed location by your snap:

1. libdrm:
    ```yaml
    parts:

>       drm-i386:

Reviewer 2: The `-i386` suffix is specific to supporting i386 on x86_64 - this should be noted in a separate section I think, pointing at the Mesa implementation, as an optional feature.

        ...
        organize:
        # Expected at /libdrm by the `graphics-core22` interface
        usr/share/libdrm: libdrm
    ```

2. drirc.d:
    ```yaml
    parts:
      dri-i386:
        organize:
        # Expected at /drirc.d by the `graphics-core22` interface
        usr/share/drirc.d: drirc.d
    ```

3. X11:
    ```yaml
    parts:
      x11-i386:
        organize:
        # Expected at /X11 by the `graphics-core22` interface
        usr/share/X11: X11
    ```

In addition to these paths, the snap **MUST** provide a script at the location
`bin/graphics-core22-provider-wrapper`.
This script is invoked with `graphics-core22-provider-wrapper $EXECUTABLE $ARGS...`.
This script does the following:
1. It exports any environment variables that are required for the binary 
to find and use the vendor drivers
2. It then executes the provided `$EXECUTABLE` with `$ARGS…`.

> An example of this script for the nvidia-core22 project is as follows: 

Reviewer 2: ```suggestion
An example of this script for the mesa-core22 project is as follows: 
```

```sh
# https://github.com/canonical/mesa-core22/blob/main/scripts/bin/graphics-core22-provider-wrapper.in

#!/bin/bash
set -euo pipefail

# Find the parent directory of the wrapper script. This will be
# the path of the provider snap in the client. Your files with
# be found at a path relative to this path.
SELF="$( cd -- "$(dirname "$0")/.." ; pwd -P )/usr"

# mesa-core22 supports multiarch (eg: i386 and amd64 on an amd64 system, or armhf and arm64 on an arm64 system).
# To do this mesa-core22 needs to supply multiple versions of the libraries, in different paths
# These paths are architecture-specific, so are filled in at snap build time
# @ARCH_TRIPLETS@ will be replaced with “i386-linux-gnu x86_64-linux-gnu” when built on amd64
# and “armhf-linux-gnu arm64-linux-gnu” when built on arm64
ARCH_TRIPLETS=( @ARCH_TRIPLETS@ )

# TODO: Ask Saviq about this one
if [ "$SNAP_ARCH" == "amd64" ]; then
  GCONV_PATH=${GCONV_PATH:+$GCONV_PATH:}${SELF}/lib/i386-linux-gnu/gconv

> fi

Author: @Saviq: Why are we doing this here? Chris and I had very little understanding of this part

Reviewer 2: This is to support i386 runtimes on x86_64 - one of the use cases of this whole thing was games, many of which are i386.

Reviewer 2: I think this example should be simplified to only deal with a single architecture, and a separate section should cover the optional support for other architectures, pointing at the mesa-core22 implementation for reference.

Author: So now I just provide the architecture from the snap arch triplet. Let me know if that will suffice. `i386` is mentioned separately

`graphics-core22-provider-wrapper $EXECUTABLE $ARGS...`. This script does the 
following:
1. It exports any environment variables that are required for the binary 
to find and use the vendor drivers
2. It then executes the provided `$EXECUTABLE` with `$ARGS…`.

An example of a script that provides `mesa-core22` is as follows:

```sh
# Dervied from:
# https://github.com/canonical/mesa-core22/blob/main/scripts/bin/graphics-core22-provider-wrapper.in

#!/bin/bash
set -euo pipefail

# Find the parent directory of the wrapper script. This will be
# the path of the provider snap in the client. Your files with
# be found at a path relative to this path.
SELF="$( cd -- "$(dirname "$0")/.." ; pwd -P )/usr"

# mesa-core22 supports multiarch (eg: i386 and amd64 on an amd64 system, or armhf and arm64 on an arm64 system).
# To do this mesa-core22 needs to supply multiple versions of the libraries, in different paths
# These paths are architecture-specific, so are filled in at snap build time
# @ARCH_TRIPLETS@ will be replaced with “i386-linux-gnu x86_64-linux-gnu” when built on amd64
# and “armhf-linux-gnu arm64-linux-gnu” when built on arm64

> ARCH_TRIPLETS=( @ARCH_TRIPLETS@ )

Author: True!

  LD_LIBRARY_PATH=${LD_LIBRARY_PATH}:/var/lib/snapd/lib/gl
fi

if [ -d "/var/lib/snapd/lib/glvnd/egl_vendor.d" ]; then
  # This needs to be prepended, as glvnd goes depth-first on these
  __EGL_VENDOR_LIBRARY_DIRS=/var/lib/snapd/lib/glvnd/egl_vendor.d:${__EGL_VENDOR_LIBRARY_DIRS}
fi

if [ -d "/var/lib/snapd/lib/vulkan/icd.d" ]; then
  XDG_DATA_DIRS=${XDG_DATA_DIRS}:/var/lib/snapd/lib
fi

if [ -d "/var/lib/snapd/lib/gl/vdpau" ]; then
  LD_LIBRARY_PATH=${LD_LIBRARY_PATH}:/var/lib/snapd/lib/gl/vdpau
fi

export LD_LIBRARY_PATH
# Mesa-specific environment variable pointing to directory containing $VENDOR_dri.so files
export LIBGL_DRIVERS_PATH
export LIBVA_DRIVERS_PATH
# glvnd-specific environment variable pointing to directory containing ICD descriptor files
export __EGL_VENDOR_LIBRARY_DIRS
# EGL external platform interface specific environment
export __EGL_EXTERNAL_PLATFORM_CONFIG_DIRS
export VK_LAYER_PATH
export XDG_DATA_DIRS

exec "$@"
```

This script has exported the proper paths via environment variables so that
any snap that connects to our provider snap can properly resolve these
libraries.

> ### i386 Support
> If you also need to provide your snap for `i386`, you will need to provide
> a few extra things in the `graphics-core22-provider-wrapper`.
> An example from `mesa-core22` adds these lines:
> 

Reviewer: It's highly likely that the consumers of this guide are going to be ARM-based vendors, so i386 support isn't likely to be very interesting to them. Maybe this instead?
```suggestion
### Multiarch Support
If the vendor drivers provide libraries for multiple sub-architectures (for example, 32-bit and 64-bit ARM) your provider snap can supply both. In this case, you will need to include all provided arch triples in the `ARCH_TRIPLETS` array in the above script (for example: `ARCH_TRIPLETS=( armhf-linux-gnu arm64-linux-gnu )`), and will need to set a few more environment variables in the `graphics-core22-provider-wrapper`.
An example from `mesa-core22` (which provides both `x86_64-linux-gnu` and `i386-linux-gnu`) adds these lines:

```

```sh
# /bin/graphics-core22-provider-wrapper

...
if [ "$SNAP_ARCH" == "amd64" ]; then
  GCONV_PATH=${GCONV_PATH:+$GCONV_PATH:}${SELF}/lib/i386-linux-gnu/gconv
fi

if [ -d "/var/lib/snapd/lib/gl32" ] && [[ ! ${LD_LIBRARY_PATH} =~ (^|:)/var/lib/snapd/lib/gl32(:|$) ]]; then
  LD_LIBRARY_PATH=${LD_LIBRARY_PATH}:/var/lib/snapd/lib/gl32
fi

if [ -d "/var/lib/snapd/lib/gl32/vdpau" ]; then
  LD_LIBRARY_PATH=${LD_LIBRARY_PATH}:/var/lib/snapd/lib/gl32/vdpau
fi

...

[ -z ${GCONV_PATH+x} ] || export GCONV_PATH
```

The `snapcraft.yaml` file will also need to be updated to point at the 32 bit paths.
You may find an example of using `i386` in 
[mesa-core22](https://github.com/canonical/mesa-core22/blob/9d64e64fb06f7052f9e6f8a8899cac763f5fad7e/snap/snapcraft.yaml#L176).

### How to supply the vendor libraries
The following example supplies drivers from `libmali`. Using a parameter for
`make`, the libraries will be installed to `$SNAPCRAFT_PART_INSTALL/usr/lib`.
These libraries do not need to be in a particular path.

```yaml
parts:
  ...
  vendor-drivers:
  plugin: make
  source: libmali
  make-parameters:

>     - LIBDIR=$SNAPCRAFT_PART_INSTALL/usr/lib

Reviewer: ```suggestion
    - LIBDIR=$SNAPCRAFT_PART_INSTALL/usr/lib/$CRAFT_ARCH_TRIPLET_BUILD_FOR
```

This will put it into `/usr/lib/arm64-linux-gnu`, which is where the wrapper script we demonstrate will look for it.

Ubuntu 22.04 repositories. Particularly, the core `libwayland` libraries may introduce new
features since 22.04 that vendor binaries might require. This will manifest as missing
symbol errors at runtime.

Generally, new version requirements for libraries can be handled by adding a snapcraft
part building the required library version. An example of this for `libwayland` is:

```yaml
parts:
 ...
 wayland:
   plugin: meson
   meson-parameters:
     - -Ddocumentation=false
     - -Dtests=false
     - --prefix=$SNAPCRAFT_PART_INSTALL/usr
   build-packages:
     - libxml2-dev
     - libffi-dev
     - libexpat1-dev
     - pkg-config
   override-build: |
     snapcraftctl build
   source: https://gitlab.freedesktop.org/wayland/wayland.git
   source-type: git
   source-tag: 1.21.0
   prime:
     - usr/lib
```

## How to test a `graphics-core22` provider snap
Now we have our `graphics-core22` provider snap built. However, we don't yet know if it works
properly. Testing this snap is our next task.

Assuming that you have already installed the `graphics-test-tools` snap,
connect the snap that you've built to `graphics-test-tools`:

```shell

> sudo snap connect graphics-test-tools:graphics-core22 <your-snap>:graphics-core22

Author: :+1:
